[Tensor] Detect more cases of symbolic sizes/strides

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123696

Previously, we'd just check `has_symbolic_sizes_strides()` to know whether a tensor has symbolic sizes or strides; if does, we skip some profiler logic. But sometimes `has_symbolic_sizes_strides()` returns false, but we do actually have symbolic sizes or strides.

So in this change, we add `may_have_symbolic_sizes_strides()` - which should never return false if the tensor has symbolic sizes and strides

Why not change `has_symbolic_sizes_strides()`? It seems like there's preexisting logic that assumes that "if has_symbolic_sizes_strides(), then we can assume that this tensor is guaranteed to have symbolic sizes or strides". In this case, we have python-implemented sizes or strides, which should follow a different code path.

Differential Revision: [D55947660](https://our.internmc.facebook.com/intern/diff/D55947660/)

davidberard98 committed:
[Tensor] Detect more cases of symbolic sizes/strides

Previously, we'd just check `has_symbolic_sizes_strides()` to know whether a tensor has symbolic sizes or strides; if does, we skip some profiler logic. But sometimes `has_symbolic_sizes_strides()` returns false, but we do actually have symbolic sizes or strides.

So in this change, we add `may_have_symbolic_sizes_strides()` - which should never return false if the tensor has symbolic sizes and strides

Why not change `has_symbolic_sizes_strides()`? It seems like there's preexisting logic that assumes that "if has_symbolic_sizes_strides(), then we can assume that this tensor is guaranteed to have symbolic sizes or strides". In this case, we have python-implemented sizes or strides, which should follow a different code path.

Differential Revision: [D55947660](https://our.internmc.facebook.com/intern/diff/D55947660/)

[ghstack-poisoned]

davidberard98 committed:
Update on "[Tensor] Detect more cases of symbolic sizes/strides"


Previously, we'd just check `has_symbolic_sizes_strides()` to know whether a tensor has symbolic sizes or strides; if does, we skip some profiler logic. But sometimes `has_symbolic_sizes_strides()` returns false, but we do actually have symbolic sizes or strides.

So in this change, we add `may_have_symbolic_sizes_strides()` - which should never return false if the tensor has symbolic sizes and strides

Why not change `has_symbolic_sizes_strides()`? It seems like there's preexisting logic that assumes that "if has_symbolic_sizes_strides(), then we can assume that this tensor is guaranteed to have symbolic sizes or strides". In this case, we have python-implemented sizes or strides, which should follow a different code path.

Differential Revision: [D55947660](https://our.internmc.facebook.com/intern/diff/D55947660/)

[ghstack-poisoned]

aaronenyeshi commented on c10/core/TensorImpl.h:
```diff
@@ -2439,6 +2439,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     return has_symbolic_sizes_strides_;
   }
 
+  bool may_have_symbolic_sizes_strides() const {
```
> So in this change, we add may_have_symbolic_sizes_strides() - which should never return false if the tensor has symbolic sizes and strides

If this is strictly never, should it be named more strictly?

davidberard98 replied:
Hmm, in my opinion this naming more sense, e.g.:

`tensor.definitely_has_symbolic_sizes_strides()` -> if it returns true, we know that this tensor definitely has symbolic sizes/strides

`tensor.may_have_symbolic_sizes_strides()` -> if this returns true, then maybe the tensor has symbolic sizes/strides; but if this returns false, then we knwo this tensor definitely doesn't have symbolic sizes/strides.

I guess we could call it `tensor.definitely_does_not_have_symbolic_sizes_strides()` but that's a bit long...

soulitzer replied:
I would agree that if we're saying we have 3 states now: definitely yes, maybe, definitely no, it would make sense for the existing `has_symbolic_sizes_strides`  to be renamed to reflect what it is doing.

davidberard98 replied:
ok, named it "does_not_have_symbolic_sizes_strides" with a comment explaining (and inverted the return value)

aaronenyeshi reviewed (approved):
Looks great! Thanks for the fix and unit test!

davidberard98 committed:
Update on "[Tensor] Detect more cases of symbolic sizes/strides"


Previously, we'd just check `has_symbolic_sizes_strides()` to know whether a tensor has symbolic sizes or strides; if does, we skip some profiler logic. But sometimes `has_symbolic_sizes_strides()` returns false, but we do actually have symbolic sizes or strides.

So in this change, we add `may_have_symbolic_sizes_strides()` - which should never return false if the tensor has symbolic sizes and strides

Why not change `has_symbolic_sizes_strides()`? It seems like there's preexisting logic that assumes that "if has_symbolic_sizes_strides(), then we can assume that this tensor is guaranteed to have symbolic sizes or strides". In this case, we have python-implemented sizes or strides, which should follow a different code path.

Differential Revision: [D55947660](https://our.internmc.facebook.com/intern/diff/D55947660/)

[ghstack-poisoned]

davidberard98 committed:
Update on "[Tensor] Detect more cases of symbolic sizes/strides"


Previously, we'd just check `has_symbolic_sizes_strides()` to know whether a tensor has symbolic sizes or strides; if does, we skip some profiler logic. But sometimes `has_symbolic_sizes_strides()` returns false, but we do actually have symbolic sizes or strides.

So in this change, we add `may_have_symbolic_sizes_strides()` - which should never return false if the tensor has symbolic sizes and strides

Why not change `has_symbolic_sizes_strides()`? It seems like there's preexisting logic that assumes that "if has_symbolic_sizes_strides(), then we can assume that this tensor is guaranteed to have symbolic sizes or strides". In this case, we have python-implemented sizes or strides, which should follow a different code path.

Differential Revision: [D55947660](https://our.internmc.facebook.com/intern/diff/D55947660/)

[ghstack-poisoned]

aaronenyeshi reviewed (approved):
LGTM, thanks for updating the name and adding comments!

soulitzer commented on c10/core/TensorImpl.h:
```diff
@@ -2435,10 +2435,24 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  // possible that has_symbolic_sizes_strides() returns false, but we do
+  // not have symbolic sizes/strides. This exists for the case of
+  // Nested Tensor python subclass, where the sizes are implemented in python
+  // (TODO: clean this up and just implement sizes in nested tensor without a
```
The naming still feels a little confusing to me since `has_symbolic_sizes_strides() != !does_not_have_symbolic_sizes_strides()` when the naming suggests that they should be inverses?

If we wanted to do something temporary to block NT (e.g. we were planning to remove this) and avoid renaming all callsites of has_symbolic_sizes_strides, maybe its just easiest to directly modify the callsite in `execution_trace_observer.cpp` to also check `matches_policy(SizesStridesPolicy::CustomStrides)`?

davidberard98 replied:
sorry, didn't see this after it was edited I think. I think `matches_policy` is `protected` so it's not accessible outside of `TensorImpl`. I guess we could also make that public, but I'm not sure if that's too different from this change?

soulitzer reviewed (approved):
Thanks for the quick fix! 

My only comment is that the naming is still confusing imo and we're leaning on the comment to explain everything.

soulitzer commented:
I don't think there's a better name unless we want to update all existing callsites too (which seems not worth it if this 3-state thing is temporary anyway)

my suggestion was just to avoid adding a new API altogether if we were just going to remove it later

(OR if we believe that subclasses in general can be in the situation where a instance can "sometimes" have symbolic shape, I think the naming from your comment here https://github.com/pytorch/pytorch/pull/123696#discussion_r1559950417 makes sense - though I cannot think of any examples of such subclasses)
